x/perpetual/keeper/begin_blocker.go

```diff
@@ -28,46 +28,55 @@ func (k Keeper) BeginBlocker(ctx sdk.Context) {
 				ctx.Logger().Error(err.Error())
 			}
-			err = k.UpdateFundingRate(ctx, &pool)
-			if err != nil {
-				ctx.Logger().Error(err.Error())
-			}
-
-			// account custody from long position
-			totalCustodyLong := sdk.ZeroInt()
-			for _, asset := range pool.PoolAssetsLong {
-				totalCustodyLong = totalCustodyLong.Add(asset.Custody)
-			}
 
-			// account custody from short position
-			totalCustodyShort := sdk.ZeroInt()
-			for _, asset := range pool.PoolAssetsShort {
-				totalCustodyShort = totalCustodyShort.Add(asset.Custody)
-			}
-
-			fundingAmountLong := types.CalcTakeAmount(totalCustodyLong, pool.FundingRate)
-			fundingAmountShort := sdk.ZeroInt()
-
-			fundingRateLong := pool.FundingRate
-			fundingRateShort := sdk.ZeroDec()
+			fundingRateLong, fundingRateShort := k.ComputeFundingRate(ctx, pool)
 
-			// if funding rate is negative, collect from short position
-			if pool.FundingRate.IsNegative() {
-				fundingAmountShort = types.CalcTakeAmount(totalCustodyShort, pool.FundingRate)
-				fundingAmountLong = sdk.ZeroInt()
-
-				fundingRateLong = sdk.ZeroDec()
-				fundingRateShort = pool.FundingRate
-			}
+			// TODO: Change this calculation based on time, going to it in another PR
 			k.SetFundingRate(ctx, uint64(ctx.BlockHeight()), pool.AmmPoolId, types.FundingRateBlock{
 				FundingRate:        pool.FundingRate,
 				BlockHeight:        ctx.BlockHeight(),
-				FundingAmountShort: fundingAmountShort,
-				FundingAmountLong:  fundingAmountLong,
+				FundingAmountShort: sdk.ZeroInt(),
+				FundingAmountLong:  sdk.ZeroInt(),
 				FundingRateLong:    fundingRateLong,
 				FundingRateShort:   fundingRateShort,
 			})
 		}
 		k.SetPool(ctx, pool)
 	}
 }
+
+func (k Keeper) ComputeFundingRate(ctx sdk.Context, pool types.Pool) (sdk.Dec, sdk.Dec) {
+	// Custody amount for long is trading asset -
+	// Liability amount for short is trading asset
+	// popular_rate = fixed_rate * abs(Custody-Liability) / (Custody+Liability)
+	totalCustodyLong := sdk.ZeroInt()
+	for _, asset := range pool.PoolAssetsLong {
+		totalCustodyLong = totalCustodyLong.Add(asset.Custody)
+	}
+
+	totalLiabilitiesShort := sdk.ZeroInt()
+	for _, asset := range pool.PoolAssetsShort {
+		totalLiabilitiesShort = totalLiabilitiesShort.Add(asset.Liabilities)
+	}
+
+	fixedRate := k.GetParams(ctx).FixedFundingRate
+	if totalCustodyLong.GT(totalLiabilitiesShort) {
+		// long is popular
+		// long pays short
+		if totalLiabilitiesShort.IsZero() {
+			return sdk.ZeroDec(), sdk.ZeroDec()
+		} else {
+			long := sdk.NewDecFromInt(totalCustodyLong.Sub(totalLiabilitiesShort)).Quo(sdk.NewDecFromInt(totalCustodyLong.Add(totalLiabilitiesShort)))
+			return long.Mul(fixedRate), sdk.ZeroDec()
+		}
+	} else {
+		// short is popular
+		// short pays long
+		if totalCustodyLong.IsZero() {
+			return sdk.ZeroDec(), sdk.ZeroDec()
+		} else {
+			short := sdk.NewDecFromInt(totalLiabilitiesShort.Sub(totalCustodyLong)).Quo(sdk.NewDecFromInt(totalCustodyLong.Add(totalLiabilitiesShort)))
```

Comment:
Also, instead of using sdk.NewDecFromInt can directly call ToLegacyDec(). More functional programming approach and reduces my work in v0.50.9 upgrade (function  NewDecFromInt got changed)